ffi_internals/src/consumer/consumer_fn.rs

```diff
@@ -0,0 +1,120 @@
+            ffi_parameters = self.ffi_calling_arguments(),
+            close_conversion = close_conversion,
+        )
+    }
+
+    /// Generates the contents of a consumer extension for this function, extending the original
+    /// type with the behaviors described by `self`. This is primarily for use with
+    /// `ffi_derive::expose_fn`, where we want to generate an FFI and consumer for a standalone
+    /// function.
+    ///
+    #[must_use]
+    pub fn generate_consumer_extension(&self, consumer_type: &str, module_name: &Ident) -> String {
+        // Include the keyword `static` if this function doesn't take a receiver.
+        let static_keyword = if self.receiver == FnReceiver::None {
+            "static"
+        } else {
+            ""
+        };
+        let (return_conversion, close_conversion, return_sig) =
+            self.return_type.as_ref().map_or_else(
+                || (String::new(), String::new(), String::new()),
+                crate::type_ffi::TypeFFI::consumer_return_type_components,
+            );
+
+        let extension = format!(
+            "
+extension {consumer_type} {{
+
+    {static_keyword} func {consumer_fn_name}({consumer_parameters}) {return_sig} {{
+        {return_conversion}{ffi_fn_name}({ffi_parameters}){close_conversion}
+    }}
+
+}}
+",
+            static_keyword = static_keyword,
+            consumer_type = consumer_type,
+            consumer_fn_name = self.fn_name.to_string().to_mixed_case(),
+            consumer_parameters = self.consumer_parameters(),
+            return_sig = return_sig,
+            return_conversion = return_conversion,
+            ffi_fn_name = self.ffi_fn_name(module_name).to_string(),
+            ffi_parameters = self.ffi_calling_arguments(),
+            close_conversion = close_conversion,
+        );
+
+        [super::header_and_imports(&[]), extension].join("")
+    }
+
+    fn consumer_parameters(&self) -> String {
+        self.parameters
+            .iter()
+            .map(|arg| {
+                format!(
+                    "{}: {}",
+                    arg.name.to_string(),
+                    arg.native_type_data.consumer_type(None)
+                )
+            })
+            .collect::<Vec<String>>()
+            .join(", ")
+    }
+
+    fn ffi_calling_arguments(&self) -> String {
+        let mut parameters: Vec<String> = self
+            .parameters
+            .iter()
+            .map(|arg| {
+                let clone_or_borrow = if arg.native_type_data.argument_borrows_supported() {
+                    "borrowReference"
+                } else {
+                    "clone"
+                };
+                format!("{}.{}()", arg.name.to_string(), clone_or_borrow)
+            })
+            .collect();
+        if self.receiver != FnReceiver::None {
+            let receiver_arg = "pointer".to_string();
+            parameters.insert(0, receiver_arg);
+        }
+        parameters.join(", ")
+    }
+}
```

Comment:
it just occurred to me, while wondering if I'd have any issues jumping in to this crate(s) to make a fix or add a feature... should we have more unit tests for these things?

Reply:
Yeah, `ffi_consumer` is horribly untested (at least without running the result against an Xcode project)... I'll add some tests to this crate for verifying the output of individual functions, and see about adding some kind of integration test that just makes sure the Swift output compiles. https://github.com/agrian-inc/ffi_common/issues/15